src/machines/signal_external_state_machine.rs

```diff
@@ -57,14 +129,214 @@ impl SignalExternalCommandRecorded {
     }
     pub(super) fn on_signal_external_workflow_execution_failed(
         self,
+        cause: SignalExternalWorkflowExecutionFailedCause,
     ) -> SignalExternalMachineTransition<Failed> {
-        unimplemented!()
+        TransitionResult::commands(vec![SignalExternalCommand::Failed(cause)])
     }
 }
 
 #[derive(Default, Clone)]
 pub(super) struct Signaled {}
+
+impl TryFrom<CommandType> for SignalExternalMachineEvents {
+    type Error = ();
+
+    fn try_from(c: CommandType) -> Result<Self, Self::Error> {
+        Ok(match c {
+            CommandType::SignalExternalWorkflowExecution => {
+                Self::CommandSignalExternalWorkflowExecution
+            }
+            _ => return Err(()),
+        })
+    }
+}
+impl TryFrom<HistoryEvent> for SignalExternalMachineEvents {
+    type Error = WFMachinesError;
+
+    fn try_from(e: HistoryEvent) -> Result<Self, Self::Error> {
+        Ok(match e.event_type() {
+            EventType::ExternalWorkflowExecutionSignaled => Self::ExternalWorkflowExecutionSignaled,
+            EventType::SignalExternalWorkflowExecutionInitiated => {
+                Self::SignalExternalWorkflowExecutionInitiated
+            }
+            EventType::SignalExternalWorkflowExecutionFailed => {
+                if let Some(
+                    history_event::Attributes::SignalExternalWorkflowExecutionFailedEventAttributes(
+                        attrs,
+                    ),
+                ) = e.attributes
+                {
+                    Self::SignalExternalWorkflowExecutionFailed(attrs.cause())
+                } else {
+                    return Err(WFMachinesError::Fatal(format!(
+                        "Signal workflow failed attributes were unset: {}",
+                        e
+                    )));
+                }
+            }
+            _ => {
+                return Err(WFMachinesError::Nondeterminism(format!(
+                    "Signal external WF machine does not handle this event: {}",
+                    e
+                )))
+            }
+        })
+    }
+}
+
+impl WFMachinesAdapter for SignalExternalMachine {
+    fn adapt_response(
+        &self,
+        my_command: Self::Command,
+        _event_info: Option<EventInfo>,
+    ) -> Result<Vec<MachineResponse>, WFMachinesError> {
+        Ok(match my_command {
+            SignalExternalCommand::Signaled => {
+                vec![ResolveSignalExternalWorkflow {
+                    seq: self.shared_state.seq,
+                    failure: None,
+                }
+                .into()]
+            }
+            SignalExternalCommand::Failed(f) => {
+                vec![ResolveSignalExternalWorkflow {
+                    seq: self.shared_state.seq,
+                    // TODO: Create new failure type upstream for this
+                    failure: Some(Failure {
+                        message: "Unable to signal external workflow because it was not found"
```

Comment:
The only failure cause in `SignalExternalWorkflowExecutionFailedCause` is not found, so for now that's really all we can say

Reply:
I think I made a similar assumption somewhere in the child WF machine.
Can we match the failure cause reported by the server in case another cause is added?